### PR TITLE
Iteration2 develop parallel

### PR DESCRIPTION
**Pull request by sanketsmali03**
@samiran-bs  
you can ignore the following during your review 
- Jenkins Allure Report issue : **Fixed**
- Tags  for gradle is not working   **Fixed**
- add reference variable for repeated values in gradle & maven file **Fixed**
- Specifying scenario name which is parameterized through commandline arguments doesn't work eg: Login with given <username> user is not working **Fixed**
- Tab 8 Spacing indentation **Fixed**
- Update readme ( formatting , adding dataproviderthreadcount ) **Fixed**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,171 @@
+buildscript {
+    ext {
+        TESTNG_VERSION = '7.3.0'
+        CUCUMBER_VERSION = '6.7.0'
+       	APPLICATION_URL = 'http://bstackdemo.com/'
```
Comment (abhi-singhs) [inline build.gradle]: fix indentation.

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+
+	<properties>
+		<maven.compiler.source>1.8</maven.compiler.source>
+		<maven.compiler.target>1.8</maven.compiler.target>
+		<maven-surefire-plugin.version>2.22.2</maven-surefire-plugin.version>
+		<selenium-java.version>3.141.59</selenium-java.version>
+		<browserstack-local-java.version>1.0.6</browserstack-local-java.version>
+		 <allure-cucumber6-jvm.version>2.14.0</allure-cucumber6-jvm.version>
+        <allure.version>2.10.0</allure.version>
```
Comment (abhi-singhs) [inline pom.xml]: fix indentation.

File: src/test/java/com/browserstack/examples/core/ManagedWebDriver.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package com.browserstack.examples.core;
+
+import org.openqa.selenium.WebDriver;
+
+import com.browserstack.examples.core.config.Platform;
+
+/**
```
Comment (abhi-singhs) [inline src/test/java/com/browserstack/examples/core/ManagedWebDriver.java]: Remove auto-generated comments.

**Review by abhi-singhs — COMMENTED** [review]
Let us follow the 8 tab-size spacing for all the files for all indentations.

File: src/test/java/com/browserstack/examples/tests/RunWebDriverCucumberTests.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	public void tearDownClass() {
+		if (testNGCucumberRunner == null) {
+			return;
+		}
+		testNGCucumberRunner.finish();
+	}
+
+	@AfterSuite
+	public void ReportGeneration() {
+		//Utility.moveFolder();
```
Comment (abhi-singhs) [inline src/test/java/com/browserstack/examples/tests/RunWebDriverCucumberTests.java]: Use for this?
Function not implemented.

File: src/test/resources/config/bs.json [hunk 5]
```diff
@@ -0,0 +1,184 @@
+{
```
Comment (abhi-singhs) [inline src/test/resources/config/bs.json]: Let us have either json or yaml configuration.

File: build.gradle [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+version '1.0-SNAPSHOT'
+repositories {
+    mavenCentral()
+    maven {
+        url "https://plugins.gradle.org/m2/"
+    }
+
+}
+dependencies {
+    testImplementation group:'org.seleniumhq.selenium', name:'selenium-java', version:"$SELENIUM_VERSION"
```
Comment (abhi-singhs) [inline build.gradle:38]: Commons Lang3 missing from dependencies.
Gradle profiles not working.

File: src/test/resources/config/bs.json (hunk 5 above)
Comment (sanketsmali03) [reply src/test/resources/config/bs.json]: removed the json file

File: src/test/java/com/browserstack/examples/tests/RunWebDriverCucumberTests.java (hunk 4 above)
Comment (sanketsmali03) [reply src/test/java/com/browserstack/examples/tests/RunWebDriverCucumberTests.java]: Removed this function now , it was earlier used for moving allure generated files to target folder

File: src/test/java/com/browserstack/examples/core/ManagedWebDriver.java (hunk 3 above)
Comment (sanketsmali03) [reply src/test/java/com/browserstack/examples/core/ManagedWebDriver.java]: Removed the auto-generated comments

File: build.gradle (hunk 6 above)
Comment (sanketsmali03) [reply build.gradle:38]: Fixed this

File: pom.xml (hunk 2 above)
Comment (sanketsmali03) [reply pom.xml]: Indentation is fixed now

File: build.gradle (hunk 1 above)
Comment (sanketsmali03) [reply build.gradle]: Indentation is fixed now

**Review by abhi-singhs — APPROVED** [review]

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,20 @@
+target
+build
+.gradle
+.idea
+*.iml
+/test-output/*
+/target/*
+/allure-results/*
+/Screenshots/*
```
Comment (samiran-bs) [inline .gitignore]: `/Screenshots/*`, `/test-output/*` directories are created by which lib? or do we have a custom code written for  generating those folders?

File: JenkinsFile [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            git branch: 'iteration2_develop_parallel', changelog: false, poll: false, url: 'https://github.com/browserstack/browserstack-examples-cucumber-testng'
+        }
+    }
+
+    stage('Start Local') {
+        if ("${params.TEST_TYPE}".contains('local')) {
+            dir('app') {
+                git branch: 'master', changelog: false, poll: false, url: 'https://github.com/browserstack/browserstack-demo-app'
+                sh label: '', returnStatus: true, script: '''#!/bin/bash -l
+                cd browserstack - demo - app
```
Comment (samiran-bs) [inline JenkinsFile]: spaces in the folder name is needed?

File: build.gradle [hunk 9]
```diff
@@ -0,0 +1,177 @@
+buildscript {
+    ext {
+        TESTNG_VERSION = '7.3.0'
```
Comment (samiran-bs) [inline build.gradle]: lets test once with 7.4.0. if that works .. lets keep that since its released in feb 2021

File: build.gradle [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        CUCUMBER_VERSION = '6.7.0'
```
Comment (samiran-bs) [inline build.gradle]: let's use 6.10.x unless that causes errors.

File: build.gradle [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+       	APPLICATION_URL = 'http://bstackdemo.com/'
```
Comment (samiran-bs) [inline build.gradle]: lets explicitly use `https`

File: build.gradle [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+       	APPLICATION_LOCAL_URL = 'http://localhost:3000/'
+        ALLURE_VERSION = "2.14.0"
+        ALLURE_GRADLE_VERSION = "2.8.1"
+        SELENIUM_VERSION = "3.141.59"
+        BROWSERSTACK_LOCAL_VERSION="1.0.6"
+        APACHE_COMMON_VERSION = "3.11"
+        LOGBACK_VERSION = "1.2.3"
+        TESTNG_VERSION="7.4.0"
+        CUCUMBER_VERSION = "6.10.4"
```
Comment (samiran-bs) [inline build.gradle]: duplicate variable defined?

File: build.gradle [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    ext {
+        TESTNG_VERSION = '7.3.0'
+        CUCUMBER_VERSION = '6.7.0'
+       	APPLICATION_URL = 'http://bstackdemo.com/'
+       	APPLICATION_LOCAL_URL = 'http://localhost:3000/'
+        ALLURE_VERSION = "2.14.0"
+        ALLURE_GRADLE_VERSION = "2.8.1"
+        SELENIUM_VERSION = "3.141.59"
+        BROWSERSTACK_LOCAL_VERSION="1.0.6"
+        APACHE_COMMON_VERSION = "3.11"
```
Comment (samiran-bs) [inline build.gradle]: 3.12.x march 2021 version

File: build.gradle [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+        LOGBACK_VERSION = "1.2.3"
+        TESTNG_VERSION="7.4.0"
```
Comment (samiran-bs) [inline build.gradle]: duplicate variable TESTNG_VERSION

File: build.gradle [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+        TESTNG_VERSION = '7.3.0'
+        CUCUMBER_VERSION = '6.7.0'
+       	APPLICATION_URL = 'http://bstackdemo.com/'
+       	APPLICATION_LOCAL_URL = 'http://localhost:3000/'
+        ALLURE_VERSION = "2.14.0"
+        ALLURE_GRADLE_VERSION = "2.8.1"
+        SELENIUM_VERSION = "3.141.59"
+        BROWSERSTACK_LOCAL_VERSION="1.0.6"
+        APACHE_COMMON_VERSION = "3.11"
+        LOGBACK_VERSION = "1.2.3"
```
Comment (samiran-bs) [inline build.gradle]: not used in this file.

File: build.gradle [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+        TESTNG_VERSION="7.4.0"
+        CUCUMBER_VERSION = "6.10.4"
+        JACKSON_DATABIND_VERSION = "2.12.2"
+        APACHE_COMMON_VERSION = "3.12.0"
```
Comment (samiran-bs) [inline build.gradle:12]: duplicate variable defined?

File: pom.xml [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+		<maven.compiler.source>1.8</maven.compiler.source>
+		<maven.compiler.target>1.8</maven.compiler.target>
+		<maven-surefire-plugin.version>2.22.2</maven-surefire-plugin.version>
+		<selenium-java.version>3.141.59</selenium-java.version>
+		<browserstack-local-java.version>1.0.6</browserstack-local-java.version>
+		<allure-cucumber6-jvm.version>2.14.0</allure-cucumber6-jvm.version>
+		<allure.version>2.10.0</allure.version>
+		<cucumber.version>6.9.1</cucumber.version>
+		<testng.version>7.4.0</testng.version>
+		<cucumber.version>6.10.4</cucumber.version>
```
Comment (samiran-bs) [inline pom.xml]: duplicate `cucumber.version` variable in the same file. pls define the latest

File: README.md [hunk 18]
```diff
@@ -1,2 +1,515 @@
...
+  ```
+
+  Gradle:
+  ```sh
+  ./gradlew bstack-local -Dtest-name="<Test scenario name>"
+  ```
+
+  where,  the argument 'test-name' can be any Cucumber scenario name configured in this repository.
+
+  E.g. "Login as username", "Login as Locked User", "Offers for mumbai geo-location" or any of the other test scenario names, as outlined in [About the tests in this repository](#About-the-tests-in-this-repository) section.
```
Comment (samiran-bs) [inline README.md]: Test names are not consistent with `About the tests in this repository` section

File: README.md [hunk 19]
```diff
@@ -1,2 +1,517 @@
...
+
+- BrowserStack
+
+  - Maven:
+
+    [pom.xml](pom.xml)  parallel-count = 5
+  
+  - Command Line args:
+    
+    -Ddataproviderthreadcount= Thread-count
```
Comment (samiran-bs) [inline README.md]: `<Thread-count>`

File: src/test/resources/capabilities-bstack-parallel-browsers.yml [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+#    - name: Android_Samsung Galaxy S21_11.0
+#      os: Android
+#      os_version: '11.0'
+#      device: Samsung Galaxy S21
+#      real_mobile: true
+#    - name: iOS_iPhone XS_13.0
+#      os: iPhone
+#      os_version: '13.0'
+#      device: iPhone XS
+#      real_mobile: true
```
Comment (samiran-bs) [inline src/test/resources/capabilities-bstack-parallel-browsers.yml]: uncomment or remove, as applicable.

File: src/test/resources/capabilities-bstack-single.yml [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+  localTunnel:
+    enable: false
+  common_capabilities:
+    project: BrowserStack Demo Repository
+    buildPrefix: browserstack-examples-cucumber-testng
+    capabilities:
+      browserstack.debug: true
+      browserstack.networkLogs: true
+      browserstack.console: debug
+#      browserstack.idleTimeout: 300
```
Comment (samiran-bs) [inline src/test/resources/capabilities-bstack-single.yml]: remove. this flag shouldn't be added unless there is a relevant use case.

File: README.md [hunk 22]
```diff
@@ -1,2 +1,517 @@
...
+    ```
+
+## About the tests in this repository
+
+This repository contains the following Selenium tests:
+
+| Module   | Test name                          | Description |
+  | ---   | ---                                   | --- |
+| E2E      | End to End Scenario                | This test scenario verifies successful product purchase lifecycle end-to-end and is also the default test executed in all the single test run profiles. |
+| Login    | Login with given username user         | This test verifies the login workflow with different types of valid login users. |
```
Comment (samiran-bs) [inline README.md]: are we able to call `Scenario Outline: Login with given <username> user` by passing `Login with given username user` in the terminal? just checking if this works, haven't pulled the code on my machine

**Review by samiran-bs — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 7 above)
Comment (sanketsmali03) [reply .gitignore]: There was a separate code earlierfor generating screenshots but now its not i will remove them

File: README.md (hunk 22 above)
Comment (sanketsmali03) [reply README.md]: This is the format -  **Login with given 'fav_user' user** 

 I have already mentioned in the readme example

File: .gitignore (hunk 7 above)
Comment (sanketsmali03) [reply .gitignore]: Fixed

File: JenkinsFile (hunk 8 above)
Comment (sanketsmali03) [reply JenkinsFile]: Fixed

File: build.gradle (hunk 9 above)
Comment (sanketsmali03) [reply build.gradle]: Fixed

File: build.gradle (hunk 10 above)
Comment (sanketsmali03) [reply build.gradle]: I am not able to set 6.10.x , it is giving error during project build , updated to latest version : Fixed

File: build.gradle (hunk 11 above)
Comment (sanketsmali03) [reply build.gradle]: Fixed

File: build.gradle (hunk 12 above)
Comment (sanketsmali03) [reply build.gradle]: Fixed

File: build.gradle (hunk 13 above)
Comment (sanketsmali03) [reply build.gradle]: Fixed

File: build.gradle (hunk 14 above)
Comment (sanketsmali03) [reply build.gradle]: Fixed

File: build.gradle (hunk 15 above)
Comment (sanketsmali03) [reply build.gradle]: Fixed ( Removed the dependency )

File: build.gradle (hunk 16 above)
Comment (sanketsmali03) [reply build.gradle:12]: Fixed

File: pom.xml (hunk 17 above)
Comment (sanketsmali03) [reply pom.xml]: Fixed

File: README.md (hunk 19 above)
Comment (sanketsmali03) [reply README.md]: Fixed

File: src/test/resources/capabilities-bstack-parallel-browsers.yml (hunk 20 above)
Comment (sanketsmali03) [reply src/test/resources/capabilities-bstack-parallel-browsers.yml]: Fixed ( uncommented the caps  )

File: README.md (hunk 18 above)
Comment (sanketsmali03) [reply README.md]: Fixed

File: src/test/resources/capabilities-bstack-single.yml (hunk 21 above)
Comment (sanketsmali03) [reply src/test/resources/capabilities-bstack-single.yml]: Fixed

**Review by samiran-bs — APPROVED** [review]

**Pull request closed without merging**